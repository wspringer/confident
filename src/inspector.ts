import {set, snakeCase} from 'lodash';
import * as yup from 'yup';
import {ObjectShape} from 'yup/lib/object';
import {SchemaObjectDescription, SchemaFieldDescription} from 'yup/lib/schema';

export const inspect = <T1, T2 extends yup.ObjectSchema<ObjectShape, T1>>(
  schema: T2
) => {
  return analyzeDesc(schema.describe());
};

export interface Descriptor {
  envVar: string;
  set: (obj: any, value: string) => void;
  desc?: string;
  path: string;
}

const analyzeDesc = (
  desc: SchemaObjectDescription,
  acc: Descriptor[] = [],
  prefix?: string,
  prefixPath?: string
) => {
  if (desc.type !== 'object') acc;
  for (const key in desc.fields) {
    const segment = snakeCase(key).toUpperCase();
    const variable = prefix ? `${prefix}_${segment}` : segment;
    const path = prefixPath ? `${prefixPath}.${key}` : key;
    if (desc.fields[key].type === 'array') {
      console.warn(`Arrays are not supported (${variable})`);
    }
    if (desc.fields[key].type !== 'object') {
      acc.push({
        envVar: variable,
        set: (obj: any, value: any) => {
          set(obj, path, value);
        },
        desc: (desc.fields[key] as any)?.meta?.desc,
        path,
      });
    } else {
      const updated = desc.fields[key] as SchemaObjectDescription;
      analyzeDesc(updated, acc, variable, path);
    }
  }
  return acc;
};
