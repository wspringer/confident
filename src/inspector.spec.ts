import * as yup from 'yup';
import {inspect} from './inspector';

describe('meta', () => {
  it('should allow you to get the meta data', () => {
    const schema = yup.object().shape({
      port: yup.number().integer().meta({desc: 'The port number'}),
      name: yup.string().meta({desc: 'The name'}),
    });
    const descriptors = inspect(schema);
    expect(descriptors).toMatchSnapshot();
  });

  it('should allow you to set object data', () => {
    const schema = yup.object().shape({
      port: yup.number().integer().meta({desc: 'The port number'}),
      name: yup.string().meta({desc: 'The name'}),
    });
    const descriptors = inspect(schema);
    const obj = {};
    descriptors[0].set(obj, '12');
    expect(obj).toMatchSnapshot();
  });

  it('should allow you to set nested object data', () => {
    const schema = yup.object().shape({
      sf: yup.object().shape({
        port: yup.number().integer().meta({desc: 'The port number'}),
      }),
    });
    const descriptors = inspect(schema);
    const obj = {};
    descriptors[0].set(obj, '12');
    expect(descriptors[0].envVar).toEqual('SF_PORT');
    expect(obj).toMatchSnapshot();
  });
});
