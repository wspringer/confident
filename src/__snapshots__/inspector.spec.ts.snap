// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`meta should allow you to get the meta data 1`] = `
Array [
  Object {
    "desc": "The port number",
    "envVar": "PORT",
    "path": "port",
    "set": [Function],
  },
  Object {
    "desc": "The name",
    "envVar": "NAME",
    "path": "name",
    "set": [Function],
  },
]
`;

exports[`meta should allow you to set nested object data 1`] = `
Object {
  "sf": Object {
    "port": "12",
  },
}
`;

exports[`meta should allow you to set object data 1`] = `
Object {
  "port": "12",
}
`;
